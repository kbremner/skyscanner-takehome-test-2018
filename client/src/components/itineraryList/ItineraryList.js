import React from 'react';
import { connect } from 'react-redux';
import Waypoint from 'react-waypoint';
import { BpkExtraLargeSpinner, SPINNER_TYPES } from 'bpk-component-spinner';
import { search, fetchNextPage } from '../../store/actions';
import { loading, error, itineraries, hasMorePages } from '../../store/selectors';
import PropTypes from '../../types';
import ItineraryCard from './ItineraryCard';
import ItineraryError from './ItineraryError';
import './ItineraryList.scss';

export class ItineraryList extends React.PureComponent {

  componentDidMount() {
    this.props.search();
  }

  render() {
    const { itineraries, search, fetchNextPage, loading, error, hasMorePages } = this.props;
    return (<div className="itinerary-list">
      {error && <ItineraryError
        className="itinerary-list__error"
        onRetry={search}
      />}
      {!error && ((loading || itineraries.length > 0)
        ? itineraries.map(itinerary => (
            <ItineraryCard
              className="itinerary-list__card"
              itinerary={itinerary}
              key={itinerary.bookingLink}
            />
          ))
      : <div>no results!</div>)}
      {loading && <BpkExtraLargeSpinner className="itinerary-list__spinner" type={SPINNER_TYPES.primary} />}
      {!error && !loading && hasMorePages && itineraries.length > 0 && <Waypoint
        onEnter={fetchNextPage}
        bottomOffset={-10}
      />}
    </div>);
  }
}

ItineraryList.propTypes = {
  itineraries: PropTypes.arrayOf(PropTypes.itinerary).isRequired,
  search: PropTypes.func.isRequired,
  fetchNextPage: PropTypes.func.isRequired,
  loading: PropTypes.bool.isRequired,
  error: PropTypes.bool.isRequired,
  hasMorePages: PropTypes.bool.isRequired
};

const mapStateToProps = state => {
  return {
    loading: loading(state),
    error: error(state),
    itineraries: itineraries(state),
    hasMorePages: hasMorePages(state)
  }
}

const mapDispatchToProps = dispatch => {
  return {
    search: () => dispatch(search()),
    fetchNextPage: () => dispatch(fetchNextPage())
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(ItineraryList);
