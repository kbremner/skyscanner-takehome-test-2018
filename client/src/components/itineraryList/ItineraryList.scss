@import "~bpk-mixins";

.itinerary-list {
  flex-grow: 1;
  background-color: $bpk-color-gray-100;
  overflow-y: auto;
  overflow-anchor: none;
  text-align: center;
  padding: 0 $bpk-spacing-sm;

  .itinerary-list__card {
    margin: $bpk-spacing-sm auto;
  }

  .itinerary-list__spinner {
    margin: $bpk-spacing-md;
  }
}