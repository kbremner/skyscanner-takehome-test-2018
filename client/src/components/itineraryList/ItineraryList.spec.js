import React from 'react';
import { shallow } from 'enzyme';
import { ItineraryList } from './ItineraryList';

const defaultItinerary = (id) => ({
  price: '$123.45',
  agentName: `test agent`,
  bookingLink: `http://example.com/${id}`,
  legs: [
    {
      id: '1',
      carrier: {
        name: 'easyjet',
        code: 'EZ'
      },
      duration: 210,
      numStops: 0,
      segments: [11, 12, 13, 14].map(val => ({
        origin: {
          dateTime: `${val}:00`, code: `CODE-${id}-${val}`
        },
        destination: {
          dateTime: `${val+1}:00`, code: `CODE-${id}-${val+1}`
        }
      }))
    }
  ]
});

const defaultItineraries = () => ([
  defaultItinerary('1'),
  defaultItinerary('2'),
  defaultItinerary('3'),
]);

const setup = (overrideProps = {}) => {
  const props = {
    itineraries: defaultItineraries(),
    error: false,
    loading: false,
    hasMorePages: true,
    search: jest.fn(),
    fetchNextPage: jest.fn(),
    ...overrideProps
  };
  const wrapper = shallow(<ItineraryList {...props} />);
  return { wrapper, props };
};

describe('ItineraryList', () => {
  describe('error', () => {
    it('is not rendered when error prop is false', () => {
      const { wrapper } = setup({ error: false });
      expect(wrapper.find('ItineraryError').length).toBe(0);
    });
  
    it('renders when error prop is true', () => {
      const { wrapper } = setup({ error: true });
      expect(wrapper).toMatchSnapshot();
    });
  
    it('calls provided handler when retrying', () => {
      const { wrapper, props } = setup({ error: true });
      props.search.mockClear();
      wrapper.find('ItineraryError').simulate('retry');
      expect(props.search).toHaveBeenCalledTimes(1);
    });
  });

  describe('spinner', () => {
    it('does not render when not fetching results', () => {
      const { wrapper } = setup({ loading: false });
      expect(wrapper.find('BpkExtraLargeSpinner').length).toBe(0);
    });
  
    it('renders when fetching results', () => {
      const { wrapper } = setup({ loading: true });
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('waypoint', () => {
    it('does not render when there is an error', () => {
      const { wrapper } = setup({ error: true });
      expect(wrapper.find('Waypoint').length).toBe(0);
    });
  
    it('does not render when loading', () => {
      const { wrapper } = setup({ loading: true });
      expect(wrapper.find('Waypoint').length).toBe(0);
    });
  
    it('does not render when there are no more pages', () => {
      const { wrapper } = setup({ hasMorePages: false });
      expect(wrapper.find('Waypoint').length).toBe(0);
    });
  
    it('does not render when there are no itineraries', () => {
      const { wrapper } = setup({ hasMorePages: true, itineraries: [] });
      expect(wrapper.find('Waypoint').length).toBe(0);
    });
  
    it('calls provided handler when Waypoint enters viewport', () => {
      const { wrapper, props } = setup({ hasMorePages: true });
      wrapper.find('Waypoint').simulate('enter');
      expect(props.fetchNextPage).toHaveBeenCalledTimes(1);
    });
  });

  describe('itineraries', () => {
    it('do not render when there is an error', () => {
      const { wrapper } = setup({ error: true });
      expect(wrapper.find('ItineraryCard').length).toBe(0);
    });

    it('renders empty message when there are no itineraries and not loading', () => {
      const { wrapper } = setup({ loading: false, itineraries: [] });
      expect(wrapper).toMatchSnapshot();
    });

    it('renders when there are itineraries', () => {
      const { wrapper } = setup();
      expect(wrapper).toMatchSnapshot();
    });

    it('renders when there are itineraries and is loading', () => {
      const { wrapper } = setup({ loading: true });
      expect(wrapper).toMatchSnapshot();
    });
  });

  it('calls search when mounted', () => {
    const { wrapper, props } = setup({ error: true });
    expect(props.search).toHaveBeenCalledTimes(1);
  });
});
