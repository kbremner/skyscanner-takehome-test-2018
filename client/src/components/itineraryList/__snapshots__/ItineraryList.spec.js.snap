// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItineraryList error renders when error prop is true 1`] = `
<div
  className="itinerary-list"
>
  <ItineraryError
    className="itinerary-list__error"
    onRetry={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
      }
    }
  />
</div>
`;

exports[`ItineraryList itineraries renders empty message when there are no itineraries and not loading 1`] = `
<div
  className="itinerary-list"
>
  <div>
    no results!
  </div>
</div>
`;

exports[`ItineraryList itineraries renders when there are itineraries 1`] = `
<div
  className="itinerary-list"
>
  <ItineraryCard
    className="itinerary-list__card"
    itinerary={
      Object {
        "agentName": "test agent",
        "bookingLink": "http://example.com/1",
        "legs": Array [
          Object {
            "carrier": Object {
              "code": "EZ",
              "name": "easyjet",
            },
            "duration": 210,
            "id": "1",
            "numStops": 0,
            "segments": Array [
              Object {
                "destination": Object {
                  "code": "CODE-1-12",
                  "dateTime": "12:00",
                },
                "origin": Object {
                  "code": "CODE-1-11",
                  "dateTime": "11:00",
                },
              },
              Object {
                "destination": Object {
                  "code": "CODE-1-13",
                  "dateTime": "13:00",
                },
                "origin": Object {
                  "code": "CODE-1-12",
                  "dateTime": "12:00",
                },
              },
              Object {
                "destination": Object {
                  "code": "CODE-1-14",
                  "dateTime": "14:00",
                },
                "origin": Object {
                  "code": "CODE-1-13",
                  "dateTime": "13:00",
                },
              },
              Object {
                "destination": Object {
                  "code": "CODE-1-15",
                  "dateTime": "15:00",
                },
                "origin": Object {
                  "code": "CODE-1-14",
                  "dateTime": "14:00",
                },
              },
            ],
          },
        ],
        "price": "$123.45",
      }
    }
    key="http://example.com/1"
  />
  <ItineraryCard
    className="itinerary-list__card"
    itinerary={
      Object {
        "agentName": "test agent",
        "bookingLink": "http://example.com/2",
        "legs": Array [
          Object {
            "carrier": Object {
              "code": "EZ",
              "name": "easyjet",
            },
            "duration": 210,
            "id": "1",
            "numStops": 0,
            "segments": Array [
              Object {
                "destination": Object {
                  "code": "CODE-2-12",
                  "dateTime": "12:00",
                },
                "origin": Object {
                  "code": "CODE-2-11",
                  "dateTime": "11:00",
                },
              },
              Object {
                "destination": Object {
                  "code": "CODE-2-13",
                  "dateTime": "13:00",
                },
                "origin": Object {
                  "code": "CODE-2-12",
                  "dateTime": "12:00",
                },
              },
              Object {
                "destination": Object {
                  "code": "CODE-2-14",
                  "dateTime": "14:00",
                },
                "origin": Object {
                  "code": "CODE-2-13",
                  "dateTime": "13:00",
                },
              },
              Object {
                "destination": Object {
                  "code": "CODE-2-15",
                  "dateTime": "15:00",
                },
                "origin": Object {
                  "code": "CODE-2-14",
                  "dateTime": "14:00",
                },
              },
            ],
          },
        ],
        "price": "$123.45",
      }
    }
    key="http://example.com/2"
  />
  <ItineraryCard
    className="itinerary-list__card"
    itinerary={
      Object {
        "agentName": "test agent",
        "bookingLink": "http://example.com/3",
        "legs": Array [
          Object {
            "carrier": Object {
              "code": "EZ",
              "name": "easyjet",
            },
            "duration": 210,
            "id": "1",
            "numStops": 0,
            "segments": Array [
              Object {
                "destination": Object {
                  "code": "CODE-3-12",
                  "dateTime": "12:00",
                },
                "origin": Object {
                  "code": "CODE-3-11",
                  "dateTime": "11:00",
                },
              },
              Object {
                "destination": Object {
                  "code": "CODE-3-13",
                  "dateTime": "13:00",
                },
                "origin": Object {
                  "code": "CODE-3-12",
                  "dateTime": "12:00",
                },
              },
              Object {
                "destination": Object {
                  "code": "CODE-3-14",
                  "dateTime": "14:00",
                },
                "origin": Object {
                  "code": "CODE-3-13",
                  "dateTime": "13:00",
                },
              },
              Object {
                "destination": Object {
                  "code": "CODE-3-15",
                  "dateTime": "15:00",
                },
                "origin": Object {
                  "code": "CODE-3-14",
                  "dateTime": "14:00",
                },
              },
            ],
          },
        ],
        "price": "$123.45",
      }
    }
    key="http://example.com/3"
  />
  <Waypoint
    bottomOffset={-10}
    fireOnRapidScroll={true}
    horizontal={false}
    onEnter={[MockFunction]}
    onLeave={[Function]}
    onPositionChange={[Function]}
    topOffset="0px"
  />
</div>
`;

exports[`ItineraryList itineraries renders when there are itineraries and is loading 1`] = `
<div
  className="itinerary-list"
>
  <ItineraryCard
    className="itinerary-list__card"
    itinerary={
      Object {
        "agentName": "test agent",
        "bookingLink": "http://example.com/1",
        "legs": Array [
          Object {
            "carrier": Object {
              "code": "EZ",
              "name": "easyjet",
            },
            "duration": 210,
            "id": "1",
            "numStops": 0,
            "segments": Array [
              Object {
                "destination": Object {
                  "code": "CODE-1-12",
                  "dateTime": "12:00",
                },
                "origin": Object {
                  "code": "CODE-1-11",
                  "dateTime": "11:00",
                },
              },
              Object {
                "destination": Object {
                  "code": "CODE-1-13",
                  "dateTime": "13:00",
                },
                "origin": Object {
                  "code": "CODE-1-12",
                  "dateTime": "12:00",
                },
              },
              Object {
                "destination": Object {
                  "code": "CODE-1-14",
                  "dateTime": "14:00",
                },
                "origin": Object {
                  "code": "CODE-1-13",
                  "dateTime": "13:00",
                },
              },
              Object {
                "destination": Object {
                  "code": "CODE-1-15",
                  "dateTime": "15:00",
                },
                "origin": Object {
                  "code": "CODE-1-14",
                  "dateTime": "14:00",
                },
              },
            ],
          },
        ],
        "price": "$123.45",
      }
    }
    key="http://example.com/1"
  />
  <ItineraryCard
    className="itinerary-list__card"
    itinerary={
      Object {
        "agentName": "test agent",
        "bookingLink": "http://example.com/2",
        "legs": Array [
          Object {
            "carrier": Object {
              "code": "EZ",
              "name": "easyjet",
            },
            "duration": 210,
            "id": "1",
            "numStops": 0,
            "segments": Array [
              Object {
                "destination": Object {
                  "code": "CODE-2-12",
                  "dateTime": "12:00",
                },
                "origin": Object {
                  "code": "CODE-2-11",
                  "dateTime": "11:00",
                },
              },
              Object {
                "destination": Object {
                  "code": "CODE-2-13",
                  "dateTime": "13:00",
                },
                "origin": Object {
                  "code": "CODE-2-12",
                  "dateTime": "12:00",
                },
              },
              Object {
                "destination": Object {
                  "code": "CODE-2-14",
                  "dateTime": "14:00",
                },
                "origin": Object {
                  "code": "CODE-2-13",
                  "dateTime": "13:00",
                },
              },
              Object {
                "destination": Object {
                  "code": "CODE-2-15",
                  "dateTime": "15:00",
                },
                "origin": Object {
                  "code": "CODE-2-14",
                  "dateTime": "14:00",
                },
              },
            ],
          },
        ],
        "price": "$123.45",
      }
    }
    key="http://example.com/2"
  />
  <ItineraryCard
    className="itinerary-list__card"
    itinerary={
      Object {
        "agentName": "test agent",
        "bookingLink": "http://example.com/3",
        "legs": Array [
          Object {
            "carrier": Object {
              "code": "EZ",
              "name": "easyjet",
            },
            "duration": 210,
            "id": "1",
            "numStops": 0,
            "segments": Array [
              Object {
                "destination": Object {
                  "code": "CODE-3-12",
                  "dateTime": "12:00",
                },
                "origin": Object {
                  "code": "CODE-3-11",
                  "dateTime": "11:00",
                },
              },
              Object {
                "destination": Object {
                  "code": "CODE-3-13",
                  "dateTime": "13:00",
                },
                "origin": Object {
                  "code": "CODE-3-12",
                  "dateTime": "12:00",
                },
              },
              Object {
                "destination": Object {
                  "code": "CODE-3-14",
                  "dateTime": "14:00",
                },
                "origin": Object {
                  "code": "CODE-3-13",
                  "dateTime": "13:00",
                },
              },
              Object {
                "destination": Object {
                  "code": "CODE-3-15",
                  "dateTime": "15:00",
                },
                "origin": Object {
                  "code": "CODE-3-14",
                  "dateTime": "14:00",
                },
              },
            ],
          },
        ],
        "price": "$123.45",
      }
    }
    key="http://example.com/3"
  />
  <BpkExtraLargeSpinner
    className="itinerary-list__spinner"
    type="primary"
  />
</div>
`;

exports[`ItineraryList spinner renders when fetching results 1`] = `
<div
  className="itinerary-list"
>
  <ItineraryCard
    className="itinerary-list__card"
    itinerary={
      Object {
        "agentName": "test agent",
        "bookingLink": "http://example.com/1",
        "legs": Array [
          Object {
            "carrier": Object {
              "code": "EZ",
              "name": "easyjet",
            },
            "duration": 210,
            "id": "1",
            "numStops": 0,
            "segments": Array [
              Object {
                "destination": Object {
                  "code": "CODE-1-12",
                  "dateTime": "12:00",
                },
                "origin": Object {
                  "code": "CODE-1-11",
                  "dateTime": "11:00",
                },
              },
              Object {
                "destination": Object {
                  "code": "CODE-1-13",
                  "dateTime": "13:00",
                },
                "origin": Object {
                  "code": "CODE-1-12",
                  "dateTime": "12:00",
                },
              },
              Object {
                "destination": Object {
                  "code": "CODE-1-14",
                  "dateTime": "14:00",
                },
                "origin": Object {
                  "code": "CODE-1-13",
                  "dateTime": "13:00",
                },
              },
              Object {
                "destination": Object {
                  "code": "CODE-1-15",
                  "dateTime": "15:00",
                },
                "origin": Object {
                  "code": "CODE-1-14",
                  "dateTime": "14:00",
                },
              },
            ],
          },
        ],
        "price": "$123.45",
      }
    }
    key="http://example.com/1"
  />
  <ItineraryCard
    className="itinerary-list__card"
    itinerary={
      Object {
        "agentName": "test agent",
        "bookingLink": "http://example.com/2",
        "legs": Array [
          Object {
            "carrier": Object {
              "code": "EZ",
              "name": "easyjet",
            },
            "duration": 210,
            "id": "1",
            "numStops": 0,
            "segments": Array [
              Object {
                "destination": Object {
                  "code": "CODE-2-12",
                  "dateTime": "12:00",
                },
                "origin": Object {
                  "code": "CODE-2-11",
                  "dateTime": "11:00",
                },
              },
              Object {
                "destination": Object {
                  "code": "CODE-2-13",
                  "dateTime": "13:00",
                },
                "origin": Object {
                  "code": "CODE-2-12",
                  "dateTime": "12:00",
                },
              },
              Object {
                "destination": Object {
                  "code": "CODE-2-14",
                  "dateTime": "14:00",
                },
                "origin": Object {
                  "code": "CODE-2-13",
                  "dateTime": "13:00",
                },
              },
              Object {
                "destination": Object {
                  "code": "CODE-2-15",
                  "dateTime": "15:00",
                },
                "origin": Object {
                  "code": "CODE-2-14",
                  "dateTime": "14:00",
                },
              },
            ],
          },
        ],
        "price": "$123.45",
      }
    }
    key="http://example.com/2"
  />
  <ItineraryCard
    className="itinerary-list__card"
    itinerary={
      Object {
        "agentName": "test agent",
        "bookingLink": "http://example.com/3",
        "legs": Array [
          Object {
            "carrier": Object {
              "code": "EZ",
              "name": "easyjet",
            },
            "duration": 210,
            "id": "1",
            "numStops": 0,
            "segments": Array [
              Object {
                "destination": Object {
                  "code": "CODE-3-12",
                  "dateTime": "12:00",
                },
                "origin": Object {
                  "code": "CODE-3-11",
                  "dateTime": "11:00",
                },
              },
              Object {
                "destination": Object {
                  "code": "CODE-3-13",
                  "dateTime": "13:00",
                },
                "origin": Object {
                  "code": "CODE-3-12",
                  "dateTime": "12:00",
                },
              },
              Object {
                "destination": Object {
                  "code": "CODE-3-14",
                  "dateTime": "14:00",
                },
                "origin": Object {
                  "code": "CODE-3-13",
                  "dateTime": "13:00",
                },
              },
              Object {
                "destination": Object {
                  "code": "CODE-3-15",
                  "dateTime": "15:00",
                },
                "origin": Object {
                  "code": "CODE-3-14",
                  "dateTime": "14:00",
                },
              },
            ],
          },
        ],
        "price": "$123.45",
      }
    }
    key="http://example.com/3"
  />
  <BpkExtraLargeSpinner
    className="itinerary-list__spinner"
    type="primary"
  />
</div>
`;
